### PR TITLE
IGNITE-14664

### PR DESCRIPTION
**Pull request by sanpwc**
Initial implementation of client side meta storage and raft related logic.

### CODE REVIEW

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Entry> get(@NotNull Key key, long revUpperBound) {
+        return metaStorageRaftGrpSvc.run(new GetCommand(key, revUpperBound));
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Map<Key, Entry>> getAll(Collection<Key> keys) {
+        List<Key> keysWithPreservedOrder = new ArrayList<>(keys);
```
Comment (agura) [inline modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java:90]: Why wee need a preserved order here and below? The order doesn't matter. We can just get `Entry.key()` from result.

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        private void stopWatch(IgniteUuid watchId) {
+            watchers.computeIfPresent(
+                watchId,
+                (k, v) -> {
+                    CompletableFuture.runAsync(v::interrupt).thenRun(() -> {
+                        try {
+                            v.cursor.close();
+                        }
+                        catch (Exception e) {
+                            e.printStackTrace();
```
Comment (agura) [inline modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java:287]: Use logging for exceptions.

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/PutCommand.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @return The key. Couldn't be .
+     */
+    public @NotNull Key key() {
+        return key;
+    }
+
+    /**
+     * @return The value.Couldn't be .
```
Comment (alievmirza) [inline modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/PutCommand.java:52]:
```suggestion
     * @return The value.Couldn't be {@code null}.
```

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/RemoveCommand.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * @param key he key. Couldn't be {@code null}.
+     */
+    public RemoveCommand(@NotNull Key key) {
+        this.key = key;
+    }
+
+    /**
+     * @return The key. Couldn't be .
+     */
```
Comment (alievmirza) [inline modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/RemoveCommand.java:40]: the same

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java (hunk 2 above)
Comment (sanpwc) [reply modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java:287]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/KeyValueStorageImpl.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+package org.apache.ignite.metastorage.common;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.ignite.internal.metastorage.common.DummyEntry;
+import org.jetbrains.annotations.NotNull;
+
+// TODO: IGNITE-14389 Tmp, should be removed.
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/KeyValueStorageImpl.java:26]: Yep, please pay attention that the only reason why `MetaStorageCommandListener` is parametrized with KeyValueStorage is testing.

File: modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/WatchEvent.java [hunk 6]
```diff
@@ -52,7 +55,21 @@ public Entry oldEntry() {
      *
      * @return New entry.
      */
-    public Entry newEntry() {
+    public @NotNull Entry newEntry() {
         return newEntry;
     }
+
+    /** {@inheritDoc} */
+    @Override public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/WatchEvent.java:63]: Fixed.

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.ignite.internal.metastorage.common.command.cursor.CursorNextCommand;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.apache.ignite.lang.IgniteLogger;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.metastorage.common.Cursor;
+import org.apache.ignite.raft.client.service.RaftGroupService;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Meta storage service side implementation of cursor.
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java:34]: Nope, we either should use MetaStorage like it's named in code, or Meta storage as a literary equivalent.
As a reference I've used javadoc in MetaStorageService.class
```
/**
 * Defines interface for access to a meta storage service.
 */
public interface MetaStorageService {
```

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/PutCommand.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+public final class PutCommand implements WriteCommand {
+    /** The key. Couldn't be {@code null}. */
+    @NotNull private final Key key;
+
+    /** The value.Couldn't be {@code null}. */
+    @NotNull private final byte[] val;
+
+    /**
+     * @param key The key. Couldn't be {@code null}.
+     * @param val The value.Couldn't be {@code null}.
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/PutCommand.java:37]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/PutCommand.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    public PutCommand(@NotNull Key key, @NotNull byte[] val) {
+        this.key = key;
+        this.val = val;
+    }
+
+    /**
+     * @return The key. Couldn't be .
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/PutCommand.java:45]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/RemoveCommand.java (hunk 4 above)
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/RemoveCommand.java:40]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/WatchExactKeysCommand.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        if (keys instanceof Serializable)
+            this.keys = keys;
+        else
+            this.keys = new ArrayList<>(keys);
+
+        this.revision = revision;
+    }
+
+    /**
+     * @return The keys collection. Couldn't be .
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/WatchExactKeysCommand.java:51]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/WatchExactKeysCommand.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public @NotNull Collection<Key> keys() {
+        return keys;
+    }
+
+    /**
+     * @return Start revision inclusive.  - all revisions.
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/WatchExactKeysCommand.java:58]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/cursor/CursorHasNextCommand.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.apache.ignite.internal.metastorage.common.command.cursor;
+
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.raft.client.ReadCommand;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Cursor hasNext command for MetaStorageCommandListener that checks whether next element is available.
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/internal/metastorage/common/command/cursor/CursorHasNextCommand.java:25]: Fixed.

File: modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/raft/MetaStorageCommandListener.java [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+import org.apache.ignite.metastorage.common.OperationTimeoutException;
+import org.apache.ignite.metastorage.common.WatchEvent;
+import org.apache.ignite.raft.client.ReadCommand;
+import org.apache.ignite.raft.client.WriteCommand;
+import org.apache.ignite.raft.client.service.CommandClosure;
+import org.apache.ignite.raft.client.service.RaftGroupCommandListener;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Meta storage command listener aka mata storage raft state machine.
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-common/src/main/java/org/apache/ignite/metastorage/common/raft/MetaStorageCommandListener.java:61]: aka removed. I don't think that we need raft state machine description here.

**Pull request closed without merging**